SBDebugger: Add new APIs `AddDestroyCallback` and `RemoveDestroyCallback`

royitaqi opened the pull request:
# Motivation

Individual callers of `SBDebugger::SetDestroyCallback()` might think that they have registered their callback and expect it to be called when the debugger is destroyed. In reality, only the last caller survives, and all previous callers are forgotten, which might be a surprise to them. Worse, if this is called in a race condition, which callback survives is less predictable, which may case confusing behavior elsewhere.

# This PR

Allows multiple destroy callbacks to be registered and all called when the debugger is destroyed.

**EDIT**: Adds two new APIs: `AddDestroyCallback()` and `ClearDestroyCallback()`. `SetDestroyCallback()` will first clear then add the given callback. Tests are added for the new APIs.

## Tests

```
bin/llvm-lit -sv ../external/llvm-project/lldb/test/API/python_api/debugger/TestDebuggerAPI.py
```

## (out-dated, see comments below) Semantic change to `SetDestroyCallback()`

~~Currently, the method overwrites the old callback with the new one. With this PR, it will NOT overwrite. Instead, it will hold on to both. Both callbacks get called during destroy.~~

~~**Risk**: Although the documentation of `SetDestroyCallback()` (see [C++](https://lldb.llvm.org/cpp_reference/classlldb_1_1SBDebugger.html#afa1649d9453a376b5c95888b5a0cb4ec) and [python](https://lldb.llvm.org/python_api/lldb.SBDebugger.html#lldb.SBDebugger.SetDestroyCallback)) doesn't really specify the behavior, there is a risk: if existing call sites rely on the "overwrite" behavior, they will be surprised because now the old callback will get called.  But as the above said, the current behavior of "overwrite" itself might be unintended, so I don't anticipate users to rely on this behavior. In short, this risk might be less of a problem if we correct it sooner rather than later (which is what this PR is trying to do).~~

## (out-dated, see comments below) Implementation

~~The implementation holds a `std::vector<std::pair<callback, baton>>`. When `SetDestroyCallback()` is called, callbacks and batons are appended to the `std::vector`. When destroy event happen, the `(callback, baton)` pairs are invoked FIFO. Finally, the `std::vector` is cleared.~~

# (out-dated, see comments below) Alternatives considered

~~Instead of changing `SetDestroyCallback()`, a new method `AddDestroyCallback()` can be added, which use the same `std::vector<std::pair<>>` implementation. Together with `ClearDestroyCallback()` (see below), they will replace and deprecate `SetDestroyCallback()`. Meanwhile, in order to be backward compatible, `SetDestroyCallback()` need to be updated to clear the `std::vector` and then add the new callback. Pros: The end state is semantically more correct. Cons: More steps to take; potentially maintaining an "incorrect" behavior (of "overwrite").~~

~~A new method `ClearDestroyCallback()` can be added. Might be unnecessary at this point, because workflows which need to set then clear callbacks may exist but shouldn't be too common at least for now. Such method can be added later when needed.~~

~~The `std::vector` may bring slight performance drawback if its implementation doesn't handle small size efficiently. However, even if that's the case, this path should be very cold (only used during init and destroy). Such performance drawback should be negligible.~~

~~A different implementation was also considered. Instead of using `std::vector`, the current `m_destroy_callback` field can be kept unchanged. When `SetDestroyCallback()` is called, a lambda function can be stored into `m_destroy_callback`. This lambda function will first call the old callback, then the new one. This way, `std::vector` is avoided. However, this implementation is more complex, thus less readable, with not much perf to gain.~~

royitaqi committed:
Allow multiple destroy callbacks in `SBDebugger::SetDestroyCallback()`

royitaqi committed:
Fix code format

JDevlieghere reviewed (changes requested):
Can you elaborate a bit more about why you want to change the behavior? Your motivation touches on the fact that it might be surprising or racy. While I think having the ability to register multiple callbacks makes sense, I'm not sure I agree with either.

 - The method is called `SetDestroyCallback` which makes it pretty clear that it's a setter, which is generally expected to change the value, rather than add to it. If the method had been called `AddDestroyCallback` on the other hand, I would expect it to add the callback rather than overwrite it. 
 - I don't think there's anything racy about a setter. Even if it were, nothing in this PR addresses the racy-ness.

The setter also makes it possible to remove a callback by overwriting it with a callback that does nothing. With the current approach that's no longer possible. I'm not sure if that's important, but something to consider. 

If the goal is to be able to specify multiple callbacks, a potential solution could be to store the callbacks in a list as you do in this PR, but add a new method `AddDestroyCallback` which appends to the list. We could keep the existing behavior for `SetDestroyCallback` by clearing the list and adding it as the first entry. That would also solve the problem of not being able to remove existing callbacks. 

Regardless of the direction, this definitely needs a corresponding test and updated documentation.

jimingham commented:
I would also find it surprising if "SetDestroyCallback" did "AddDestroyCallback".  That's not what the name says it does.  I have no problem with supporting multiple Destroy callbacks.  But I agree with Jonas, that needs to be a separate API with an appropriate name.  And also tests...

royitaqi commented:
Thank you, @JDevlieghere and @jimingham for the input.

IIUR, you are basically advocating for the following alternative approach (that was mentioned in the "Alternatives considered" section in the above):

> Instead of changing SetDestroyCallback(), a new method AddDestroyCallback() can be added, which use the same std::vector<std::pair<>> implementation. Together with ClearDestroyCallback() (see below), they will replace and deprecate SetDestroyCallback(). Meanwhile, in order to be backward compatible, SetDestroyCallback() need to be updated to clear the std::vector and then add the new callback. Pros: The end state is semantically more correct. Cons: More steps to take; potentially maintaining an "incorrect" behavior (of "overwrite").

> A new method ClearDestroyCallback() can be added. Might be unnecessary at this point, because workflows which need to set then clear callbacks may exist but shouldn't be too common at least for now. Such method can be added later when needed.


I agree that adding new methods have more clear naming and don't change the semantics of the eixsting method. I will update this PR (or, **should I create a new PR?** - Noob I am.)

royitaqi committed:
Add `AddDestroyCallback()` and `ClearDestroyCallback()`

jimingham commented:
I don't have a strong preference for new PR vrs. reuse this one.

royitaqi committed:
Reflect API changes in `SBDebugger` class and add test

royitaqi committed:
Update test

royitaqi committed:
Improve comment and Fix code format

royitaqi commented:
Hi @JDevlieghere and @jimingham,

I have updated the PR to add `AddDestroyCallback()`, `ClearDestroyCallback()`, and tests for these. Also updated `SetDestroyCallback()` to work with the new field `std::vector<std::pair<callback, baton>>`.

Hope it looks better now. LMK if anything else needs to be added/changed.

--

BTW, noob as I am, is there anything I need to do to change the PR's state "Changes requested"?  E.g. In some other diff review tools, there is a thing to flip the state into "Ready for review".

royitaqi commented:
# Tests that I did

## Manual test

```
username-mac ~/public_llvm/build % bin/lldb
(lldb) script
Python Interactive Interpreter. To exit, type 'quit()', 'exit()' or Ctrl-D.
>>> lldb.debugger.AddDestroyCallback(lambda user_id: print('foo %s' % user_id))
>>> lldb.debugger.AddDestroyCallback(lambda user_id: print('bar %s' % user_id))
>>> ^D
now exiting InteractiveConsole...
(lldb) ^D
foo 1
bar 1
username-mac ~/public_llvm/build %
```

## Run the added unit tests

username-mac ~/public_llvm/build % bin/llvm-lit -sv /Users/username/public_llvm/llvm-project/lldb/test/API/python_api/debugger/TestDebuggerAPI.py
llvm-lit: /Users/username/public_llvm/llvm-project/lldb/test/API/lit.cfg.py:175: warning: Could not set a default per-test timeout. Requires the Python psutil module but it could not be found. Try installing it via pip or via your operating system's package manager.
llvm-lit: /Users/username/public_llvm/llvm-project/lldb/test/API/lit.cfg.py:113: note: Deleting module cache at /Users/username/public_llvm/build/lldb-test-build.noindex/module-cache-lldb/lldb-api.

Testing Time: 1.97s

Total Discovered Tests: 1
  Passed: 1 (100.00%)

1 warning(s) in tests
username-mac ~/public_llvm/build %
```

jimingham commented on lldb/source/Core/Debugger.cpp:
```diff
@@ -743,10 +743,11 @@ DebuggerSP Debugger::CreateInstance(lldb::LogOutputCallback log_callback,
-  if (m_destroy_callback) {
-    m_destroy_callback(GetID(), m_destroy_callback_baton);
-    m_destroy_callback = nullptr;
+  const lldb::user_id_t user_id = GetID();
```
A destroy callback could delete another destroy callback while running.  You need to account for that.

royitaqi replied:
I don't understand.

Were you referring to the `foreach` loop below and saying that since the container can be modified during the loop, the looping needs to be changed (e.g. to use the iterator instead of the `foreach` syntax)?

If you did meant to talk about the `user_id` then I don't think I understand how the `user_id` is related, because it seems it should be constant throughout the `HandleDestroyCallback` execution.

jimingham replied:
I meant the former.  You could have added two Destroy callbacks, but then during the execution of the first one, the code figures out that it has already done the job the second one was intended to do, so the first Destroy callback removed the second one on the way out to keep the work from getting done twice.

royitaqi replied:
Added double while-loop in `HandleDestroyCallback` to handle callbacks which can be added/removed during destroy. Added test `test_HandleDestroyCallback` to validate a simple example case.

jimingham commented on lldb/source/Core/Debugger.cpp:
```diff
@@ -1425,10 +1426,19 @@ void Debugger::SetLoggingCallback(lldb::LogOutputCallback log_callback,
 
+void Debugger::AddDestroyCallback(
+    lldb_private::DebuggerDestroyCallback destroy_callback, void *baton) {
+  m_destroy_callback_and_baton.emplace_back(destroy_callback, baton);
```
You need some locking here.  Since there's an SB API and an internal way to add destroy callbacks, you can't guarantee there won't be concurrent add operations.

royitaqi replied:
Good to know about the potential concurrency. Thanks!

royitaqi replied:
Addressed in the latest version.

jimingham commented on lldb/source/Core/Debugger.cpp:
```diff
@@ -1425,10 +1426,19 @@ void Debugger::SetLoggingCallback(lldb::LogOutputCallback log_callback,
+}
+
+void Debugger::ClearDestroyCallback() {
+  m_destroy_callback_and_baton.clear();
```
Ditto here, clear and add could be being called on separate threads so this needs some locking.

royitaqi replied:
Addressed in the latest version.

royitaqi commented on lldb/include/lldb/API/SBDebugger.h:
```diff
@@ -321,9 +321,14 @@ class LLDB_API SBDebugger {
   void SetDestroyCallback(lldb::SBDebuggerDestroyCallback destroy_callback,
                           void *baton);
 
+  void ClearDestroyCallback();
```
> It seems wrong that one client can only remove its destroy callback by removing ones it didn't install.
Makes sense to me.

I think you are suggesting:
* `TToken AddDestroyCallback(callback, baton)`
* `bool RemoveDestroyCallback(TToken token)` - returns success/fail
* No/remove `ClearDestroyCallbacks` - because no one is supposed to clear

Then what should `SetDestroyCallback` be?
* It clears existing callbacks, so I think in this new world this semantics isn't allowed, similar to how `ClearDestroyCallbacks` shouldn't exist.
* ~~That is, unless, we introduce the concept of a client ID, where `SetDestroyCallback` will remove all callbacks from the specified client ID. But then I feel the introduction of the client ID opens up even more questions, and it's not an existing pattern in the Debugger class. So it's probably not a good idea.~~

royitaqi replied:
Addressed in the latest version.

jimingham commented:
For legacy reasons, I think we have to keep SetDestroyCallbacks doing what it did, but we should comment that it is deprecated and to use the Add and Remove to only affect your own callbacks.

Jim


> On Apr 24, 2024, at 6:19 PM, royitaqi ***@***.***> wrote:
> 
> 
> @royitaqi commented on this pull request.
> 
> In lldb/include/lldb/API/SBDebugger.h <https://github.com/llvm/llvm-project/pull/89868#discussion_r1578689718>:
> 
> >    void SetDestroyCallback(lldb::SBDebuggerDestroyCallback destroy_callback,
>                            void *baton);
>  
> +  void ClearDestroyCallback();
> It seems wrong that one client can only remove its destroy callback by removing ones it didn't install.
> Makes sense to me.
> 
> I think you are suggesting:
> 
> TToken AddDestroyCallback(callback, baton)
> bool RemoveDestroyCallback(TToken token) - returns success/fail
> No/remove ClearDestroyCallbacks - because no one is supposed to clear
> Then what should SetDestroyCallback be?
> 
> It clears existing callbacks, so I think in this new world this semantics isn't allowed, similar to how ClearDestroyCallbacks shouldn't exist.
> That is, unless, we introduce the concept of a client ID, where SetDestroyCallback will remove all callbacks from the specified client ID. But then I feel the introduction of the client ID opens up even more questions, and it's not an existing pattern in the Debugger class. So it's probably not a good idea.
> —
> Reply to this email directly, view it on GitHub <https://github.com/llvm/llvm-project/pull/89868#discussion_r1578689718>, or unsubscribe <https://github.com/notifications/unsubscribe-auth/ADUPVW5DTSDZRU33KPW67BTY7BK23AVCNFSM6AAAAABGWDO5TCVHI2DSMVQWIX3LMV43YUDVNRWFEZLROVSXG5CSMV3GSZLXHMZDAMRRGIYTQMRVGY>.
> You are receiving this because you were mentioned.
>

royitaqi committed:
Thread-safe Add/Remove with tokens

royitaqi committed:
Fix format

royitaqi commented:
Hi @jimingham ,

I have updated the code. Now we have `Add` and `Remove` (with tokens as you suggested). `Set` clears everything then add the given one (also returns a token).

Thread-safety is done through a `std::lock_guard<std::recursive_mutex>`. This is consistent with some of the other functions/fields.

Didn't understand one of your comment about "one destroy callback can add/remove others" and how that relates to the code. See my reply in the above.

Another thing I am not sure is, in python API test, how do I make concurrent calls to these APIs. Do you know if this should be tested (e.g. if other existing functions has similar tests)? If the answer is "yes", do you happen to know what existing test I can see as examples?


Thanks for your reviews so far. Hope this PR is getting better as we converse.

Best,
Roy

jimingham commented on lldb/include/lldb/lldb-private-types.h:
```diff
@@ -121,6 +121,7 @@ typedef struct type256 { uint64_t x[4]; } type256;
 using ValueObjectProviderTy =
     std::function<lldb::ValueObjectSP(ConstString, StackFrame *)>;
 
+typedef int DebuggerDestroyCallbackToken;
```
There's no need to make separate lldb_private & SB versions of the CallbackToken.  Since this is intended to be a public type (it shows up in the SB API) you can put it in the lldb namespace - in lldb-types.h, and then both the SB API's and the lldb_private one can use that.

jimingham commented on lldb/include/lldb/Core/Debugger.h:
```diff
@@ -568,10 +569,22 @@ class Debugger : public std::enable_shared_from_this<Debugger>,
 
-  void
+  /// DEPRECATED. Use AddDestroyCallback and RemoveDestroyCallback instead.
+  /// Clear all previously added callbacks and only add the given one.
```
I would give a reason for the deprecation, like:

DEPRECATED: We used to only support one Destroy callback.  Now that we support Add and Remove, you should only remove Destroy callbacks that you Add-ed.

jimingham commented on lldb/source/API/SBDebugger.cpp:
```diff
@@ -1686,13 +1686,33 @@ void SBDebugger::SetLoggingCallback(lldb::LogOutputCallback log_callback,
+    return m_opaque_sp->AddDestroyCallback(
+        destroy_callback, baton);
+  }
+  return -1;
```
It's better to make an LLDB_INVALID_DESTROY_TOKEN and use that than have a magical -1 floating around in code.

royitaqi replied:
Added `LLDB_INVALID_DESTROY_CALLBACK_TOKEN`.

jimingham commented on lldb/include/lldb/API/SBDebugger.h:
```diff
@@ -321,8 +321,15 @@ class LLDB_API SBDebugger {
+                     void *baton);
+
+  lldb::SBDebuggerDestroyCallbackToken
+  SetDestroyCallback(lldb::SBDebuggerDestroyCallback destroy_callback,
```
This is also deprecated.  I'd add the same comment here, and use the LLDB_DEPRECATED_IN macro to mark it formally.

jimingham replied:
Formally, returning the token from SetDestroyCallback is a reasonable addition, but it changes the signature of an extant SB API, which we don't allow - we are strict about maintaining backwards compatibility of the SB API's.  Since we're deprecating this API anyway, I don't think we need to make Set play well with the Add/Remove model.  Just leave it returning void.

royitaqi replied:
Yeah. Thanks for catching this. Reverted the API changes (both in SBDebugger and in Debugger).

jimingham commented:
> Hi @jimingham ,
> 
> I have updated the code. Now we have `Add` and `Remove` (with tokens as you suggested). `Set` clears everything then add the given one (also returns a token).
> 
> Thread-safety is done through a `std::lock_guard<std::recursive_mutex>`. This is consistent with some of the other functions/fields.
> 
> Didn't understand one of your comment about "one destroy callback can add/remove others" and how that relates to the code. See my reply in the above.
> 
> Another thing I am not sure is, in python API test, how do I make concurrent calls to these APIs. Do you know if this should be tested (e.g. if other existing functions has similar tests)? If the answer is "yes", do you happen to know what existing test I can see as examples?
> 
> Thanks for your reviews so far. Hope this PR is getting better as we converse.
> 
> Best, Roy

I don't think it's necessary to try to test concurrency.  That tends to be very hard to write a stable test that you know actually tests concurrent access.  Those end up being more trouble than they are worth.

royitaqi committed:
Public token type and void return for Set

royitaqi committed:
Handle Add/Remove during Destroy

royitaqi committed:
Add comments

bulbazord commented on lldb/include/lldb/API/SBDebugger.h:
```diff
@@ -321,13 +321,22 @@ class LLDB_API SBDebugger {
+  /// you Add-ed. Use Add and Remove instead.
+  ///
+  /// Clear all previously added callbacks and only add the given one.
   void SetDestroyCallback(lldb::SBDebuggerDestroyCallback destroy_callback,
```
There are some macros you can use in addition to the documentation: `LLDB_DEPRECATED_FIXME`. This will give compiler warnings to anyone using `SetDestroyCallback` so they can update their code.

royitaqi replied:
Thanks for the tip! Will find an example and use it.

royitaqi replied:
Added.  I assume I only need to add it to SBDebugger.h, not Debugger.h.  Hope the message/fix I wrote is also good.

royitaqi commented on lldb/source/Core/Debugger.cpp:
```diff
@@ -743,9 +743,19 @@ DebuggerSP Debugger::CreateInstance(lldb::LogOutputCallback log_callback,
+  // In case one destroy callback adds or removes other destroy callbacks
+  // which aren't taken care of in the same inner loop.
+  while (m_destroy_callback_and_baton.size()) {
+    auto iter = m_destroy_callback_and_baton.begin();
+    while (iter != m_destroy_callback_and_baton.end()) {
+      // Invoke the callback and remove the entry from the map
+      const auto &callback = iter->second.first;
+      const auto &baton = iter->second.second;
+      callback(user_id, baton);
+      iter = m_destroy_callback_and_baton.erase(iter);
+    }
```
Readability discussion (ignore if you think this particular one is too small a topic):

I thought about writing the inner loop as `for (begin; end; ) { callback(); iter = map.erase(); }`. However, I felt it's basically the same as the above. So I just left it as is. LMK if you think the `for` loop will look better.

FWIW, if we do use the `for` loop, I didn't like the idea of moving the `iter = map.erase()` part into the 3rd clause of the `for` line. LMK if you think differently

bulbazord replied:
Instead of your 2 loop approach, could you not just repeatedly grab and call the first callback in the list? If new ones are added it'll be picked up, if others are removed then the next iteration won't see them.

royitaqi replied:
Ah I like it! It's a lot simpler this way. Will update.

royitaqi committed:
Add LLDB_DEPRECATED_FIXME

royitaqi commented:
Hi @jimingham and @bulbazord ,

I *think* I have addressed all your comments (thanks for those!). I have also eye-balled through my PR again. But it's possible that I missed something.

Will be great to have you taking a look when you have the time.

Thanks,
Roy

royitaqi commented:
Hi @JDevlieghere ,

The PR has changed quite a bit. If you have the time, maybe it's a good time to take another look.

Below is a TL;DR of some of the changes:
* Aadded Add/Remove
* Leave Set unchanged (but marked as deprecated)
* Added token in Add/Remove
* Handle concurrency
* Handle Add/Remove invocations from existing destroy callbacks during the destroy event

royitaqi committed:
Simplify loop logic and update comments

JDevlieghere commented on lldb/include/lldb/lldb-types.h:
```diff
@@ -62,12 +62,15 @@ typedef void *thread_arg_t;             // Host thread argument type
 typedef void *thread_result_t;          // Host thread result type
 typedef void *(*thread_func_t)(void *); // Host thread function type
 typedef int pipe_t;                     // Host pipe type
+typedef int destroy_callback_token_t;   // Debugger destroy callback token type
```
Should this be unsigned?

royitaqi replied:
I liked that the invalid token `LLDB_INVALID_DESTROY_CALLBACK_TOKEN` is -1, similar to how `std::basic_string::npos` is -1.

Technically I feel signed/unsigned are pretty much the same: they both have the same number of valid and invalid values; the positive/negative-ness of the values don't matter.

I don't have a strong opinion. Can change if you still think it will be better. In that case, I'm curious to learn your reasons - I must have missed something.

clayborg replied:
Since we are adding this to the public API, we should remove the "destroy" from this so we could re-use this for other callback tokens. So change this to:
```
typedef int callback_token_t;
```
And remove the comment since making a comment of `// Callback token type` doesn't really tell you anything more than the name already does. The reason being we don't want to have to add a ton of callback token types for every callback we might want to register in the future (`process_callback_token_t`, `thread_callback_token_t`, ...). We can just use one for all of them in the future.

JDevlieghere commented on lldb/include/lldb/Core/Debugger.h:
```diff
@@ -731,8 +747,11 @@ class Debugger : public std::enable_shared_from_this<Debugger>,
 
-  lldb_private::DebuggerDestroyCallback m_destroy_callback = nullptr;
-  void *m_destroy_callback_baton = nullptr;
+  std::recursive_mutex m_destroy_callback_mutex;
```
Does this actually **need** to be thread safe? APIs being thread safe is of course a good thing, but in its current state, there are plenty of methods in the Debugger class that are not guaranteed to be thread safe. Unless there's a need for this to be safe, I wouldn't bother with the new mutex... 

Either way this shouldn't be a _recursive_ mutex.

bulbazord replied:
Handling a callback could add or remove other callbacks, it does need to be thread safe. And as much as I dislike recursive mutex as well, is there another way you could handle those scenarios safely?

royitaqi replied:
> Handling a callback could add or remove other callbacks
IIUC, this answers the "recursive" part of @JDevlieghere 's comment.

> Does this actually need to be thread safe?
I had the same question in my mind when the thread safety requirement was raised. Now I look at it, the only other mutex in the same class is the global debugger list, which does sound important enough to not be messed up.

IMHO, the answer here depends on whether or not we already have a clear future where the class need to be thread safe.  If such future hasn't been aligned, I feel it can be unsafe for now, and we can add thread-safety as one-go once such direction has been aligned. WDYT?

JDevlieghere replied:
@bulbazord But that's only if you call `Debugger::Destroy` (and therefore run the callbacks) on another thread, right? Given nothing else in `(SB)Debugger` provides such safety guarantees (besides the global debugger list) I'd say let's not overthink this and if we want to change that, let's do it holistically for the whole class instead of piecemeal.

bulbazord replied:
@JDevlieghere Yes, I don't think we need to synchronize across threads, that's a good point. I was thinking about what if any of the callbacks add, remove, or otherwise invoke other callbacks. But that shouldn't require a mutex.

@royitaqi I think you can remove the mutex :)

jimingham replied:
That's true of running the destroy chain callbacks, but you certainly can add a callback from more than one thread at a time.  I think you need to at least protect adding.

royitaqi replied:
BTW:
> I think you need to at least protect adding.

To me, it's either none, or all.

If we decide to protect adding (multiple adds from different threads not during destroy), then we should just protect all of add/remove/destroy, because any of these operations can happen at the same time from multiple threads.

FWIW I agree with what @JDevlieghere said:

> I'd say let's not overthink this and if we want to change that, let's do it holistically for the whole class instead of piecemeal.

JDevlieghere commented on lldb/include/lldb/Core/Debugger.h:
```diff
@@ -731,8 +747,11 @@ class Debugger : public std::enable_shared_from_this<Debugger>,
-  void *m_destroy_callback_baton = nullptr;
+  std::recursive_mutex m_destroy_callback_mutex;
+  lldb::destroy_callback_token_t m_destroy_callback_next_token = 0;
+  std::unordered_map<lldb::destroy_callback_token_t,
+                     std::pair<lldb_private::DebuggerDestroyCallback, void *>>
+      m_destroy_callback_and_baton;
```
Given we expect the number of callbacks to be small, let's use a `llvm::SmallDenseMap` for this.

JDevlieghere commented on lldb/source/API/SBDebugger.cpp:
```diff
@@ -1690,11 +1690,28 @@ void SBDebugger::SetDestroyCallback(
+lldb::destroy_callback_token_t
+SBDebugger::AddDestroyCallback(lldb::SBDebuggerDestroyCallback destroy_callback,
+                               void *baton) {
+  LLDB_INSTRUMENT_VA(this, destroy_callback, baton);
```
Newline after `LLDB_INSTRUMENT`

JDevlieghere commented on lldb/source/API/SBDebugger.cpp:
```diff
@@ -1690,11 +1690,28 @@ void SBDebugger::SetDestroyCallback(
+}
+
+bool SBDebugger::RemoveDestroyCallback(lldb::destroy_callback_token_t token) {
+  LLDB_INSTRUMENT_VA(this);
```
Missing token argument and missing newline.

JDevlieghere commented on lldb/source/Core/Debugger.cpp:
```diff
@@ -743,9 +743,24 @@ DebuggerSP Debugger::CreateInstance(lldb::LogOutputCallback log_callback,
+  // is random, it's wise to not depend on the order and instead implement
+  // logic inside the callbacks to decide if their work should be skipped.
+  while (m_destroy_callback_and_baton.size()) {
+    auto iter = m_destroy_callback_and_baton.begin();
+    const auto &callback = iter->second.first;
+    const auto &baton = iter->second.second;
```
Using `auto` for the iterator is fine, but it's not obvious from context what the types are of `.first` and `.second`. The  [LLVM Coding Standards](https://llvm.org/docs/CodingStandards.html#use-auto-type-deduction-to-make-code-more-readable) have a paragraph on when to use (and no use) `auto`. A few more instances of that below.

royitaqi replied:
Thanks for pointing to the coding standards!
I think that paragraph makes sense.
Will update the code.

clayborg commented on lldb/include/lldb/API/SBDebugger.h:
```diff
@@ -321,9 +321,26 @@ class LLDB_API SBDebugger {
+  /// Add a callback for when the debugger is destroyed. Return a token, which
+  /// can be used to remove said callback. Multiple callbacks can be added by
+  /// calling this function multiple times.
+  lldb::destroy_callback_token_t
```
Do we really need a "lldb::destroy_callback_token_t" type here? I would prefer if this returns "void" and then modify `RemoveDestroyCallback(...)` to take the destroy callback to remove and include the baton so we know we are removing the right one:
```
bool RemoveDestroyCallback(lldb::SBDebuggerDestroyCallback destroy_callback, void *baton);
```

This would allow someone to install two callback using the same callback, but with different baton values for each, and then still remove the right one when calling `RemoveDestroyCallback(...)`

jimingham replied:
I don't see the difference.  You call:

auto token1 = my_debugger.AddDestroyCallback(my_only_callback, baton1);
auto token2 = my_debugger.AddDestroyCallback(my_only_callback, baton2);

Then when you go to delete the first one, you use `token1`, and the second, `token2`.  I don't see the problem deleting the right one.  The difference is just: do you want to have to keep around the callback/baton pair in order to potentially disable it; or do you want to hold onto the token you got back.  I think the latter is more convenient.

clayborg reviewed (changes requested):
I agree with all comments here. I like being able to add and remove destroy callbacks. See my inline comments for changing `AddDestroyCallback` and `RemoveDestroyCallback`.

royitaqi commented on lldb/include/lldb/API/SBDebugger.h:
```diff
@@ -321,9 +321,26 @@ class LLDB_API SBDebugger {
 
   void SetLoggingCallback(lldb::LogOutputCallback log_callback, void *baton);
 
+  /// DEPRECATED: We used to only support one Destroy callback. Now that we
+  /// support Add and Remove, you should only remove Destroy callbacks that
+  /// you Add-ed. Use Add and Remove instead.
```
Removing the public "DEPRECATED". Updating the private "DEPRECATED" according to your suggestion.

royitaqi committed:
Use SmallDenseMap, and coding standard fixes

royitaqi commented:
Gentle ping to make progress on this PR. :)

@JDevlieghere , could you see [my rely to your comment](https://github.com/llvm/llvm-project/pull/89868#discussion_r1588655287) about why I think why I think signed int makes sense for `typedef int destroy_callback_token_t`?  I have no strong opinion. If you still think it should be unsigned, I can update the PR.

@JDevlieghere and @bulbazord , about the necessity of thread safety, could you see if [my proposal](https://github.com/llvm/llvm-project/pull/89868#discussion_r1588661446) is agreeable? TL;DR: currently the class isn't thread safe; let's skip thread safety in this PR; we can add it to the class in one go when in the future the class is decided to be thread safe.

jimingham commented:
> On May 7, 2024, at 2:51 PM, royitaqi ***@***.***> wrote:
> 
> 
> @royitaqi commented on this pull request.
> 
> In lldb/include/lldb/Core/Debugger.h <https://github.com/llvm/llvm-project/pull/89868#discussion_r1593145799>:
> 
> > @@ -731,8 +747,11 @@ class Debugger : public std::enable_shared_from_this<Debugger>,
>    lldb::TargetSP m_dummy_target_sp;
>    Diagnostics::CallbackID m_diagnostics_callback_id;
>  
> -  lldb_private::DebuggerDestroyCallback m_destroy_callback = nullptr;
> -  void *m_destroy_callback_baton = nullptr;
> +  std::recursive_mutex m_destroy_callback_mutex;
> BTW:
> 
> I think you need to at least protect adding.
> 
> To me, it's either none, or all.
> 
IME you don't want to just keep adding mutexes willy-nilly, you do want to know that the resource could be contended for by operations on more than one thread. My only point was that we couldn't think of a plausible reason to lock when iterating for destroy, but it's easy to see that you need to lock adding and removing since there's nothing that keeps that from happening.

> If we decide to protect adding (multiple adds from different threads not during destroy), then we should just protect all of add/remove/destroy, because any of these operations can happen at the same time from multiple threads.
> 
> FWIW I agree with what @JDevlieghere <https://github.com/JDevlieghere> said:
> 
> I'd say let's not overthink this and if we want to change that, let's do it holistically for the whole class instead of piecemeal.
> 
I didn't really understand that statement.  Adding debuggers, or querying them is controlled by the debugger list mutex.  Manipulating the I/O handlers is governed by the io handler stack mutex, there's a lock around the interrupt requests.  Debugger doesn't seem like a class that is not locking the lists of things it controls as a general rule.

When you have a collection that can be accessed from multiple threads, adding and removing, then you really do need to protect those accesses.  I'm not sure it's good reasoning to say "That's not likely, let's wait till it crashes on someone to add the protection".

Jim

> —
> Reply to this email directly, view it on GitHub <https://github.com/llvm/llvm-project/pull/89868#discussion_r1593145799>, or unsubscribe <https://github.com/notifications/unsubscribe-auth/ADUPVWZU2E5VV5VVISC4TFTZBFEFHAVCNFSM6AAAAABGWDO5TCVHI2DSMVQWIX3LMV43YUDVNRWFEZLROVSXG5CSMV3GSZLXHMZDANBUGI2TSMBXGM>.
> You are receiving this because you were mentioned.
>

royitaqi commented:
@jimingham 

> Adding debuggers, or querying them is controlled by the debugger list mutex.  Manipulating the I/O handlers is governed by the io handler stack mutex, there's a lock around the interrupt requests.  Debugger doesn't seem like a class that is not locking the lists of things it controls as a general rule.

This is fair. I somehow missed them the last time I was searching for mutex in the same code.

--

With the above, I'm on the fence. Since I have less knowledge and context, at this point I will depend on @jimingham & @JDevlieghere you guys aligning with each other and pointing me in the right direction. LMK. Either ways it should be an easy change in the PR (since we already have thread-safety implemented).

royitaqi commented:
Gentle ping. :)

@jim

clayborg commented on lldb/source/API/SBDebugger.cpp:
```diff
@@ -1689,35 +1689,56 @@ void SBDebugger::SetLoggingCallback(lldb::LogOutputCallback log_callback,
+                               void *baton) {
+  LLDB_INSTRUMENT_VA(this, destroy_callback, baton);
+
+  if (m_opaque_sp) {
+    return m_opaque_sp->AddDestroyCallback(destroy_callback, baton);
+  }
```
remove braces for single line `if` statements per llvm coding guidelines.

clayborg commented on lldb/source/API/SBDebugger.cpp:
```diff
@@ -1689,35 +1689,56 @@ void SBDebugger::SetLoggingCallback(lldb::LogOutputCallback log_callback,
+bool SBDebugger::RemoveDestroyCallback(lldb::destroy_callback_token_t token) {
+  LLDB_INSTRUMENT_VA(this, token);
+
+  if (m_opaque_sp) {
+    return m_opaque_sp->RemoveDestroyCallback(token);
+  }
```
remove braces for single line if statements per llvm coding guidelines.

clayborg commented on lldb/include/lldb/Core/Debugger.h:
```diff
@@ -731,8 +747,11 @@ class Debugger : public std::enable_shared_from_this<Debugger>,
 
-  lldb_private::DebuggerDestroyCallback m_destroy_callback = nullptr;
-  void *m_destroy_callback_baton = nullptr;
+  std::recursive_mutex m_destroy_callback_mutex;
```
Just change this to a std::mutex?

royitaqi replied:
Recursive is necessary. This is because HandleDestroyCallbacks will invoke callbacks, which can call Add/RemoveDestoryCallback. All three functions use the same mutex, so it has to be recursive.

See more: https://github.com/llvm/llvm-project/pull/89868#discussion_r1588099079

royitaqi replied:
This is out-dated. Latest commit uses `std::mutex` and improve the loop in `HandleDestroyCallbacks` to avoid recurssive mutex.

clayborg reviewed (changes requested):
We should make this thread safe. It can only help and this isn't an API which gets called all of the time, so performance isn't an issue. A few inline comments

jeffreytan81 commented on lldb/include/lldb/API/SBDebugger.h:
```diff
@@ -194,7 +194,7 @@ class LLDB_API SBDebugger {
 
   void HandleCommand(const char *command);
-  
+
```
Remove unnecessary format change

jeffreytan81 commented on lldb/include/lldb/Core/Debugger.h:
```diff
@@ -13,6 +13,7 @@
 
 #include <memory>
 #include <optional>
+#include <unordered_map>
```
This is not used.

jeffreytan81 commented on lldb/test/API/python_api/debugger/TestDebuggerAPI.py:
```diff
@@ -161,3 +161,122 @@ def foo(dbg_id):
+
+        # Should call both `foo()` and `bar()`. Order is undermined because
+        # of the `unordered_map` in the implementation.
+        self.assertTrue(('foo', original_dbg_id) in called)
+        self.assertTrue(('bar', original_dbg_id) in called)
```
Use `self.assertIn()`

jeffreytan81 commented on lldb/source/Core/Debugger.cpp:
```diff
@@ -743,9 +743,28 @@ DebuggerSP Debugger::CreateInstance(lldb::LogOutputCallback log_callback,
+  // In the removal case, because the order of the container is random, it's
+  // wise to not depend on the order and instead implement logic inside the
+  // callbacks to decide if their work should be skipped.
+  while (m_destroy_callback_and_baton.size()) {
```
Use `m_destroy_callback_and_baton.empty()`

royitaqi replied:
Thought `.size()` is faster because it don't need a negation, i.e. compare to `!m_destroy_callback_and_baton.empty()`.

I do agree that the latter has better readability.

jeffreytan81 commented on lldb/source/Core/Debugger.cpp:
```diff
@@ -743,9 +743,28 @@ DebuggerSP Debugger::CreateInstance(lldb::LogOutputCallback log_callback,
+  // callbacks to decide if their work should be skipped.
+  while (m_destroy_callback_and_baton.size()) {
+    auto iter = m_destroy_callback_and_baton.begin();
+    const lldb::destroy_callback_token_t &token = iter->first;
+    const lldb_private::DebuggerDestroyCallback &callback = iter->second.first;
+    void *const &baton = iter->second.second;
```
I do not think you need to take reference to them. Copy by value is good to avoid indirection.

royitaqi replied:
I thought these will be inlined because they are only read once. So both (ref or copy) have no difference.

I will change them to copy, but am curious to learn the detail here.

jeffreytan81 commented on lldb/source/Core/Debugger.cpp:
```diff
@@ -743,9 +743,28 @@ DebuggerSP Debugger::CreateInstance(lldb::LogOutputCallback log_callback,
+    // Using `token` to erase, because elements may have been added/removed, and
+    // that will cause error "invalid iterator access!" if `iter` is used
+    // instead.
+    m_destroy_callback_and_baton.erase(token);
```
Any reason we remove after calling the callback?

royitaqi replied:
The callbacks can add/remove other callbacks while this loop is being executed. I.e. the content in the container can change during the loop.

To handle this, the loop always take the first element, call it, then remove it, so that the next "first element" will be a new one.

If we don't remove, I wonder how we can make sure that the callbacks which are added during the loop are executed.

jeffreytan81 replied:
The biggest concern is that, this implementation completely changes the semantics -- now, client using this API has to re-register the callbacks after destroying debugger which is weird and not expected because there is no removeCallback() called from client. For example, the following user case won't work anymore:
```
void MyDestroyCallback() {
...
}

SBDebugger::AddDestroyCallbac(MyDestroyCallback);
SBDebugger::Create();
...
SBDebugger::Destroy();

// Recreate another debugger
SBDebugger::Create();
SBDebugger::Destroy();   // Now the MyDestroyCallback won't be called even user did not call RemoveDestroyCallback() which is not expected
```

There are several ways to handle this issue without clearing `m_destroy_callback_and_baton`. One simply way is making a copy of `m_destroy_callback_and_baton`, and calling callbacks from the copy (by checking if it still exists in original `m_destroy_callback_and_baton`). And at the end, checking there is no new entries in `m_destroy_callback_and_baton`, otherwise, getting the delta of the local copy and original copy, and redo the process in a loop.

clayborg replied:
The callbacks shouldn't be removed when called, that is for sure

jeffreytan81 replied:
I was more concerned if AddDestroyCallback() is a static API but since it is an instance method, I am less concerned now.

jeffreytan81 commented on lldb/source/API/SBDebugger.cpp:
```diff
@@ -1689,35 +1689,56 @@ void SBDebugger::SetLoggingCallback(lldb::LogOutputCallback log_callback,
 }
 
 void SBDebugger::RequestInterrupt() {
   LLDB_INSTRUMENT_VA(this);
-  
+
   if (m_opaque_sp)
-    m_opaque_sp->RequestInterrupt();  
+    m_opaque_sp->RequestInterrupt();
 }
 void SBDebugger::CancelInterruptRequest()  {
   LLDB_INSTRUMENT_VA(this);
-  
+
   if (m_opaque_sp)
-    m_opaque_sp->CancelInterruptRequest();  
+    m_opaque_sp->CancelInterruptRequest();
 }
 
 bool SBDebugger::InterruptRequested()   {
   LLDB_INSTRUMENT_VA(this);
-  
+
   if (m_opaque_sp)
```
Undo all these unnecessary changes.

royitaqi replied:
Manually added them back, because `git clang-format` isn't clever enough to ignore the lines which I didn't modify.

royitaqi committed:
Address comments from Greg and Jeffrey

royitaqi committed:
Attempt to revert whitespace changes in other code

jeffreytan81 reviewed (commented):
Added another inline comment.

jeffreytan81 commented on lldb/source/API/SBDebugger.cpp:
```diff
@@ -1689,12 +1689,32 @@ void SBDebugger::SetLoggingCallback(lldb::LogOutputCallback log_callback,
   LLDB_INSTRUMENT_VA(this, destroy_callback, baton);
+
   if (m_opaque_sp) {
-    return m_opaque_sp->SetDestroyCallback(
-        destroy_callback, baton);
+    m_opaque_sp->SetDestroyCallback(destroy_callback, baton);
   }
```
Undo the unnecessary format change

royitaqi committed:
Use SmallVector and std::mutex; optimize the loop in HandleDestroyCallbacks

royitaqi committed:
Add comments and fix format

royitaqi committed:
Merge branch 'update-sdc' of https://github.com/royitaqi/llvm-project into update-sdc

royitaqi commented on lldb/include/lldb/Core/Debugger.h:
```diff
@@ -731,8 +746,11 @@ class Debugger : public std::enable_shared_from_this<Debugger>,
+  lldb::destroy_callback_token_t m_destroy_callback_next_token = 0;
+  llvm::SmallDenseMap<lldb::destroy_callback_token_t,
+                      std::pair<lldb_private::DebuggerDestroyCallback, void *>>
+      m_destroy_callback_and_baton;
```
Updated to use `SmallVector`.

clayborg commented on lldb/include/lldb/lldb-types.h:
```diff
@@ -62,12 +62,15 @@ typedef void *thread_arg_t;             // Host thread argument type
 #define LLDB_INVALID_HOST_THREAD ((lldb::thread_t)NULL)
 #define LLDB_INVALID_PIPE ((lldb::pipe_t)-1)
+#define LLDB_INVALID_DESTROY_CALLBACK_TOKEN                                    \
+  ((lldb::destroy_callback_token_t) - 1)
```
Remove "DESTROY" from bot the #define and _t type:
```
#define LLDB_INVALID_CALLBACK_TOKEN ((lldb::callback_token_t) - 1)
```

clayborg commented on lldb/include/lldb/Core/Debugger.h:
```diff
@@ -731,8 +747,12 @@ class Debugger : public std::enable_shared_from_this<Debugger>,
-  void *m_destroy_callback_baton = nullptr;
+  std::mutex m_destroy_callback_mutex;
+  lldb::destroy_callback_token_t m_destroy_callback_next_token = 0;
+  typedef std::tuple<lldb::destroy_callback_token_t,
+                     lldb_private::DebuggerDestroyCallback, void *>
```
Instead of a tuple, please create a struct type. Using tuples isn't very readable even though it works:
```
struct DestroyCallbackInfo {
  lldb::callback_token_t token;
  lldb_private::DebuggerDestroyCallback callback;
  void *baton;
};
```

royitaqi replied:
Added.

FWIW, I had to add two ctors:
- Default ctor for `DestroyCallbackInfo callback_info;`, used in `HandleDestroyCallback()`.
- A ctor which assigns 3 members for `m_destroy_callbacks.emplace_back(token, destroy_callback, baton);`, used in `AddDestroyCallback()`.

clayborg commented on lldb/include/lldb/Core/Debugger.h:
```diff
@@ -731,8 +747,12 @@ class Debugger : public std::enable_shared_from_this<Debugger>,
+  typedef std::tuple<lldb::destroy_callback_token_t,
+                     lldb_private::DebuggerDestroyCallback, void *>
+      DebuggerDestroyCallbackTuple;
+  llvm::SmallVector<DebuggerDestroyCallbackTuple, 2> m_destroy_callbacks;
```
llvm::SmallVector<DestroyCallbackInfo, 2> m_destroy_callbacks;

clayborg commented on lldb/source/Core/Debugger.cpp:
```diff
@@ -1427,8 +1440,30 @@ void Debugger::SetLoggingCallback(lldb::LogOutputCallback log_callback,
+  std::lock_guard<std::mutex> guard(m_destroy_callback_mutex);
+  for (auto it = m_destroy_callbacks.begin(); it != m_destroy_callbacks.end();
+       ++it) {
+    if (std::get<0>(*it) == token) {
```
Convert to use the new `DestroyCallbackInfo` struct members.

royitaqi committed:
A few renames; add struct DestroyCallbackInfo

clayborg reviewed (approved):
Thanks for all of the changes. Looks good.

royitaqi committed:
Merge branch 'main' into update-sdc

antmox commented:
Hi @royitaqi ,
This commit broke lldb-aarch64-windows buildbot : https://lab.llvm.org/buildbot/#/builders/219/builds/11564
Looks like callback_token_t is not defined for WIN32 in lldb-types.h
Could you please look at this ?

antmox commented:
@royitaqi @clayborg I proposed a fix here : https://github.com/llvm/llvm-project/pull/92870

royitaqi commented:
Hi @antmox,

I'm sorry that it broke the lldb-aarch64-windows build. My bad. Your fix makes sense, though I do have the same comment as @labath had, i.e. to move that typedef outside platform-specific region. From the latest comments in that PR it seems it's already got moved. If that's the case then I think we are all good.

My apologies that I didn't respond earlier - I was watching my email and the build bot (https://lab.llvm.org/buildbot/#/changes/134099), saw all green and I thought it's went well.   **Questions for my learning** (I'm noob to all these): I realize that the build bot I quoted didn't run the windows build you mentioned (lldb-aarch64-windows). This tells me that there are multiple build bots that I need to watch in the future. **Is there an exhaustive list somewhere?** Email obviously failed to inform me about the build break (i.e. I searched for "lldb-aarch64-windows" in my email and nothing except your message).

Thanks,
Roy
